presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayAllocator.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.block;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import static java.lang.String.format;
+
+@NotThreadSafe
+public class ArrayAllocator
+{
+    private static final int DEFAULT_MAX_OUTSTANDING = 100;
+
+    private final Deque<byte[]> byteArrays = new ArrayDeque<>();
+    private final Deque<short[]> shortArrays = new ArrayDeque<>();
+    private final Deque<int[]> intArrays = new ArrayDeque<>();
+    private final Deque<long[]> longArrays = new ArrayDeque<>();
+    private final Deque<boolean[]> booleanArrays = new ArrayDeque<>();
+
+    private int outstandingArrays;
+    private final int maxOutstandingArrays;
+
+    public ArrayAllocator()
+    {
+        this(DEFAULT_MAX_OUTSTANDING);
+    }
+
+    public ArrayAllocator(int maxOutstandingArrays)
+    {
+        this.maxOutstandingArrays = maxOutstandingArrays;
+    }
+
+    public byte[] getByteArray(int positionCount)
+    {
+        checkState(outstandingArrays++ < maxOutstandingArrays, "Requested too many arrays: %s", outstandingArrays);
```

Comment:
I have two questions here:
1) Do we really want to fail the query if outstandingArrays < maxOutstandingArrays? I understand we want to limit the size of the array pool, but should we fail the query if there're still available heap memory to ask for? In that case can we fall back to the new byte[] way while just consider it's outstanding?
2) if outstandingArrays < maxOutstandingArrays, it means the request would fail and no array from this pool is checked out. But outstandingArrays was anyways incremented by one from the ++.

Reply:
My thinking on 1 is that, because these allocators will be stored on the operator level, and not some globally accessible place, we just need a reasonable number here, and anything that exceeds this reasonable number is a bug which means there's some issue returning the array once finished.  So it's important to throw, otherwise we have a much greater chance of not catching the bug related to not returning the used arrays.

I can change 2 to increment only if successful, but it shouldn't matter since if it throws, it means the operator has failed and this allocator will be garbage collected.